### ISSUE TITLE
lib/perl5db.t: blocker for strict-by-default

### ISSUE THREAD

**Issue by jkeenan** [atoomic/perl#230]
As of commit ac6b7ccf73 in the alpha-dev-02-strict branch on Aug 18 2020, the file cited is failing.  Here is output from `cd t; ./perl harness -v <file>; cd -`:
```
# ../lib/perl5db.t

# Failed test 2 - [perl \#66110] at ../lib/perl5db.t line 95
#      got 'Global symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 6.\nGlobal symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 12.\nGlobal symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 15.\nGlobal symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 18.\nGlobal symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 21.\nGlobal symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 25.\nGlobal symbol \"$all_ok\" requires explicit package name (did you forget to declare \"my $all_ok\"?) at ../lib/perl5db/t/rt-66110 line 33.\nExecution of ../lib/perl5db/t/rt-66110 aborted due to compilation errors.\n at ../lib/perl5db/t/rt-66110 line 36.\n'
# expected /(?^:\bAll tests successful\.$)/
# Failed test 41 - n steps over lvalue subs at ../lib/perl5db.t line 1042
#      got 'Global symbol \"$x\" requires explicit package name (did you forget to declare \"my $x\"?) at ../lib/perl5db/t/lsub-n line 2.\nGlobal symbol \"$x\" requires explicit package name (did you forget to declare \"my $x\"?) at ../lib/perl5db/t/lsub-n line 4.\nGlobal symbol \"$x\" requires explicit package name (did you forget to declare \"my $x\"?) at ../lib/perl5db/t/lsub-n line 6.\nExecution of ../lib/perl5db/t/lsub-n aborted due to compilation errors.\n at ../lib/perl5db/t/lsub-n line 6.\n<>\n<>\n'
# expected /(?^:<1>\n<11>\n)/
# Failed test 44 - l statement is working properly (test No. 1). at ../lib/perl5db.t line 1113
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-2)  DB<1> l\n10376==>\t    \"Debugged program terminated.  Use \'q\' to quit or \'R\' to restart.\";\n10377 \t}\n10378 \t\n10379 \tpackage DB;    # Do not trace this 1; below!\nauto(-1)  DB<1> q\n'
# expected /(?^msx:
#             ^1==>\s+\$x\ =\ 1;\n
#             2:\s+print\ "1\\n";\n
#             3\s*\n
#             4:\s+\$x\ =\ 2;\n
#             5:\s+print\ "2\\n";\n
#         )/
# Failed test 45 - l followed by l and then followed by - at ../lib/perl5db.t line 1167
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-7)  DB<1> l\n10376==>\t    \"Debugged program terminated.  Use \'q\' to quit or \'R\' to restart.\";\n10377 \t}\n10378 \t\n10379 \tpackage DB;    # Do not trace this 1; below!\nauto(-6)  DB<1> # After l 1\n\nauto(-5)  DB<2> l\n10379 \tpackage DB;    # Do not trace this 1; below!\nauto(-4)  DB<2> # After l 2\n\nauto(-3)  DB<3> -\n10368 \t10369 \t10370 \t10371 \t10372 \t10373 \t10374 \t\n10375 \tsub at_exit {\n10376==>\t    \"Debugged program terminated.  Use \'q\' to quit or \'R\' to restart.\";\n10377 \t}\nauto(-2)  DB<4> # After -\n\nauto(-1)  DB<5> q\n'
# expected /(?^msx:
#             ^(?^msx:
#         1==>\s+\$x\ =\ 1;\n
#         2:\s+print\ "1\\n";\n
#         3\s*\n
#         4:\s+\$x\ =\ 2;\n
#         5:\s+print\ "2\\n";\n
#         6\s*\n
#         7:\s+\$x\ =\ 3;\n
#         8:\s+print\ "3\\n";\n
#         9\s*\n
#         10:\s+\$x\ =\ 4;\n
#     )
#             [^\n]*?DB<\d+>\ \#\ After\ l\ 1\n
#             [\ \t]*\n
#             [^\n]*?DB<\d+>\ l\s*\n
#             (?^msx:
#         11:\s+print\ "4\\n";\n
#         12\s*\n
#         13:\s+\$x\ =\ 5;\n
#         14:\s+print\ "5\\n";\n
#         15\s*\n
#         16:\s+\$x\ =\ 6;\n
#         17:\s+print\ "6\\n";\n
#         18\s*\n
#         19:\s+\$x\ =\ 7;\n
#         20:\s+print\ "7\\n";\n
#     )
#             [^\n]*?DB<\d+>\ \#\ After\ l\ 2\n
#             [\ \t]*\n
#             [^\n]*?DB<\d+>\ -\s*\n
#             (?^msx:
#         1==>\s+\$x\ =\ 1;\n
#         2:\s+print\ "1\\n";\n
#         3\s*\n
#         4:\s+\$x\ =\ 2;\n
#         5:\s+print\ "2\\n";\n
#         6\s*\n
#         7:\s+\$x\ =\ 3;\n
#         8:\s+print\ "3\\n";\n
#         9\s*\n
#         10:\s+\$x\ =\ 4;\n
#     )
#             [^\n]*?DB<\d+>\ \#\ After\ -\n
#         )/
# Failed test 95 - Test the = (command alias) command. at ../lib/perl5db.t line 2154
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-4)  DB<1> = quit q\nquit\t= q\nauto(-3)  DB<2> = foobar l\nfoobar\t= l\nauto(-2)  DB<3> foobar\n10376==>\t    \"Debugged program terminated.  Use \'q\' to quit or \'R\' to restart.\";\n10377 \t}\n10378 \t\n10379 \tpackage DB;    # Do not trace this 1; below!\nauto(-1)  DB<4> quit\n'
# expected /(?^msx:
#             ^1==>\s+\$x\ =\ 1;\n
#             2:\s+print\ "1\\n";\n
#             3\s*\n
#             4:\s+\$x\ =\ 2;\n
#             5:\s+print\ "2\\n";\n
#         )/
# Failed test 124 - RT 120174: p command can be invoked without space after 'p' at ../lib/perl5db.t line 2833
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-4)  DB<1> b 2\nLine 2 not breakable.\nauto(-3)  DB<2> c\nUse \'q\' to quit or \'R\' to restart.  \'h q\' for details.\nauto(-2)  DB<2> p@abc\n\nauto(-1)  DB<3> q\n'
# expected /(?^:1234)/
# Failed test 125 - RT 120174: x command can be invoked without space after 'x' before array at ../lib/perl5db.t line 2854
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-4)  DB<1> b 2\nLine 2 not breakable.\nauto(-3)  DB<2> c\nUse \'q\' to quit or \'R\' to restart.  \'h q\' for details.\nauto(-2)  DB<2> x@abc\n  empty array\nauto(-1)  DB<3> q\n'
# expected /(?^ms:0\s+1\n1\s+2\n2\s+3\n3\s+4)/
# Failed test 126 - RT 120174: x command can be invoked without space after 'x' before array ref at ../lib/perl5db.t line 2875
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-4)  DB<1> b 2\nLine 2 not breakable.\nauto(-3)  DB<2> c\nUse \'q\' to quit or \'R\' to restart.  \'h q\' for details.\nauto(-2)  DB<2> x\\@abc\n0  ARRAY(0x558689446ab0)\n     empty array\nauto(-1)  DB<3> q\n'
# expected /(?^ms:\s+0\s+1\n\s+1\s+2\n\s+2\s+3\n\s+3\s+4)/
# Failed test 127 - RT 120174: x command can be invoked without space after 'x' before hash ref at ../lib/perl5db.t line 2896
#      got '\nLoading DB routines from perl5db.pl version 1.57\nEditor support available.\n\nEnter h or \'h h\' for help, or \'man perldebug\' for more help.\n\nDebugged program terminated.  Use \e[1mq\e[0m to quit or \e[1mR\e[0m to restart,\nuse \e[1mo\e[0m \e[4minhibit_exit\e[24m to avoid stopping after program termination,\n\e[1mh q\e[0m, \e[1mh R\e[0m or \e[1mh o\e[0m to get additional info.\nauto(-4)  DB<1> b 4\nLine 4 not breakable.\nauto(-3)  DB<2> c\nUse \'q\' to quit or \'R\' to restart.  \'h q\' for details.\nauto(-2)  DB<2> x\\%xyz\n0  HASH(0x560dc246fea0)\n     empty hash\nauto(-1)  DB<3> q\n'
# expected /(?^ms:\s+'alpha'\s+=>\s+'beta'\n\s+'gamma'\s+=>\s+'delta')/
../lib/perl5db.t .. 
ok 1 - proxy constant subroutines
not ok 2 - [perl \#66110]
ok 3 - [perl \#116769] frame=2 does not crash debugger, exit == 0
ok 4 - [perl \#116769] code is run
ok 5 - [perl \#116771] autotrace does not crash debugger, exit == 0
ok 6 - [perl \#116771] code is run
ok 7 - [perl \#41461] frame=2 noTTY does not crash debugger, exit == 0
ok 8 - [perl \#41461] code is run
ok 9 - The ${main::_<filename} variable in the debugger was not destroyed
ok 10 - no strict 'vars' in evaluated lines.
ok 11 - lvalue subs work in the debugger
ok 12 - there are no undefined values in the symbol table
ok 13 # skip This perl has threads, skipping non-threaded debugger tests
ok 14 - there are no undefined values in the symbol table when running with thread support
ok 15 - [perl \#61222]
ok 16 - [perl \#104168] - level 2 appears
ok 17 - [perl \#104168] - no 'baz'
ok 18 - taint
ok 19 - Can set breakpoint in a line in the middle of the file.
ok 20 - Can set breakpoint in a line.
ok 21 - Can set breakpoint in a line.
ok 22 - Can set breakpoint in a line.
ok 23 - Can set breakpoint in a line.
ok 24 - Prompt should display the first line of code.
ok 25 - Restart and delete all breakpoints work properly.
ok 26 - 'c line_num' is working properly.
ok 27 - 'b .' is working correctly.
ok 28 - Prompt should display the line of code inside a subroutine.
ok 29 - p command works.
ok 30 - x command test.
ok 31 - x command test with '@_'.
ok 32 - Mutating '@_'.
ok 33 - x after AutoTrace=1 command is working.
ok 34 - T command test.
ok 35 - test for s - single step
ok 36 - 'b .' is working correctly.
ok 37 - Test for the s command.
ok 38 - uncalled_subroutine was called after s EXPR()
ok 39 - uncalled_subroutine was called after n EXPR()
ok 40 - b subroutine works fine
not ok 41 - n steps over lvalue subs
ok 42 - M (module list) works fine
ok 43 - r statement is working properly.
not ok 44 - l statement is working properly (test No. 1).
not ok 45 - l followed by l and then followed by -
ok 46 - l subroutine_name
ok 47 - Test the "." command
ok 48 - f command is working.
ok 49 - /pat/ command is working and found a match.
ok 50 - ?pat? command is working and found a match.
ok 51 - L command is listing breakpoints
ok 52 - L command is listing watch expressions
ok 53 - L command is not listing deleted watch expressions
ok 54 - L command is listing actions and breakpoints
ok 55 - S command - 1
ok 56 - S command with regex
ok 57 - S command with negative regex
ok 58 - S command with negative regex - what it still matches
ok 59 - a command is working
ok 60 - a command with no line number is working
ok 61 - A command (for removing actions) is working
ok 62 - 'A *' command (for removing all actions) is working
ok 63 - w command - watchpoint changed
ok 64 - w command - correct output from IDX
ok 65 - W command - watchpoint was deleted
ok 66 - W command - stopped at end.
ok 67 - "W *" command - watchpoint was deleted
ok 68 - "W *" command - stopped at end.
ok 69 - "o" command (without arguments) displays warnLevel
ok 70 - "o" command (without arguments) displays signalLevel
ok 71 - "o" command (without arguments) displays dieLevel
ok 72 - "o" command (without arguments) displays hashDepth
ok 73 - "o" query command does not display warnLevel
ok 74 - "o" query command displays signalLevel
ok 75 - "o" query command does not display dieLevel
ok 76 - "o" query command displays hashDepth
ok 77 - o set command works
ok 78 - o set command - hashDepth
ok 79 - Test < and < ? commands - contents.
ok 80 - Test < and < ? commands - output.
ok 81 - Test the '< *' command.
ok 82 - Test > and > ? commands - contents.
ok 83 - Test > and > ? commands - output.
ok 84 - Test the '> *' command.
ok 85 - Test < and > commands. 
ok 86 - Test the pre-prompt debugger commands
ok 87 - Test the { * command
ok 88 - { * test - output is OK.
ok 89 - Test the ! command (along with l 3-5)
ok 90 - Test the ! -n command (along with l)
ok 91 - Test the source command (along with l)
ok 92 - Test the source command inside a typeahead
ok 93 - Test the H -num command
ok 94 - Test the H command (without a number.)
not ok 95 - Test the = (command alias) command.
ok 96 - Test m for main - 1
ok 97 - Test m for main - 2
ok 98 - Test m for obj - 1
ok 99 - Test m for obj - 1
ok 100 - Test M
ok 101 - Test the o recallCommand option
ok 102 - Test the o dieLevel option
ok 103 - Test the o warnLevel option
ok 104 - Test the t command (without a number.)
ok 105 - Test the o AutoTrace command
ok 106 - Test the t command with function calls.
ok 107 - Test the o AutoTrace command with function calls.
ok 108 - Test the final "Debugged program terminated" message.
ok 109 - Test the o inhibit_exit=0 command.
ok 110 - Test o PrintRet=1
ok 111 - Test o PrintRet=0
ok 112 - Test o PrintRet=1 in list context
ok 113 - Test o PrintRet=0 in list context
ok 114 - Test o PrintRet=1 in void context
ok 115 - Test o PrintRet=0 in void context
ok 116 - Test o PrintRet=0 in void context
ok 117 - Test t expr
ok 118 - Test w for lexical values.
ok 119 - Test that the debugger chdirs to the initial directory after a restart.
ok 120 - perldoc command works fine
ok 121 - Test that the a command does not emit warnings on program exit.
ok 122 - Tried to connect.
ok 123 - Can quit from the debugger after a wrong RemotePort
not ok 124 - RT 120174: p command can be invoked without space after 'p'
not ok 125 - RT 120174: x command can be invoked without space after 'x' before array
not ok 126 - RT 120174: x command can be invoked without space after 'x' before array ref
not ok 127 - RT 120174: x command can be invoked without space after 'x' before hash ref
ok 128 - mro needs to be loaded
ok 129 - check for reasonable result
ok 130 - [perl \#124203] the thread ran
ok 131 - [perl \#124203] debugger didn't deadlock
ok 132 - [perl \#124203] the thread ran (lvalue)
ok 133 - [perl \#124203] debugger didn't deadlock (lvalue)
1..133
Failed 9/133 subtests 
	(less 1 skipped subtest: 123 okay)

Test Summary Report
-------------------
../lib/perl5db.t (Wstat: 0 Tests: 133 Failed: 9)
  Failed tests:  2, 41, 44-45, 95, 124-127
Files=1, Tests=133,  8 wallclock secs ( 0.03 usr  0.01 sys +  6.51 cusr  0.55 csys =  7.10 CPU)
Result: FAIL
```

@atoomic @toddr 

Thank you very much.
Jim Keenan

**Comment by toddr** [comment]
This isn't a great one for me to work on. I'd be looking at @rjbs if you could get him to take a look.

**Comment by rjbs** [comment]
I can look, but I predict it will be a nightmare.  But maybe not.

**Comment by jkeenan** [comment]
> I can look, but I predict it will be a nightmare. But maybe not.

Not a nightmare.  Just required lots of staring at it.

This should be fixed by commits commit ef0410a531a8e32c8d75657de021873f41a8d3db  and fa4e44d18e0fa0e1bbc44f97320a5ca59a5e3e05.

Thank you very much.
Jim Keenan

**Comment by jkeenan** [comment]
> > I can look, but I predict it will be a nightmare. But maybe not.
> 
> Not a nightmare. Just required lots of staring at it.
> 
> This should be fixed by commits commit [ef0410a](https://github.com/atoomic/perl/commit/ef0410a531a8e32c8d75657de021873f41a8d3db) and [fa4e44d](https://github.com/atoomic/perl/commit/fa4e44d18e0fa0e1bbc44f97320a5ca59a5e3e05).
> 

Hearing no objection in two days, closing.

Thank you very much.
Jim Keenan
